### PR TITLE
[SPARK-28323][SQL][Python] PythonUDF should be able to use in join condition

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

There is a bug in `ExtractPythonUDFs` that produces wrong result attributes. It causes a failure when using `PythonUDF`s among multiple child plans, e.g., join. An example is using `PythonUDF`s in join condition.

```python                                                                
>>> left = spark.createDataFrame([Row(a=1, a1=1, a2=1), Row(a=2, a1=2, a2=2)])                                                                                                                                                                                                                      
>>> right = spark.createDataFrame([Row(b=1, b1=1, b2=1), Row(b=1, b1=3, b2=1)])                                                                       
>>> f = udf(lambda a: a, IntegerType())                                                                                                               
>>> df = left.join(right, [f("a") == f("b"), left.a1 == right.b1])                                                
>>> df.collect()                                                                                                                                      
19/07/10 12:20:49 ERROR Executor: Exception in task 5.0 in stage 0.0 (TID 5)                                                                                                                                                 
java.lang.ArrayIndexOutOfBoundsException: 1                                                                                                           
        at org.apache.spark.sql.catalyst.expressions.GenericInternalRow.genericGet(rows.scala:201)                                                    
        at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.getAs(rows.scala:35)               
        at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.isNullAt(rows.scala:36)               
        at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.isNullAt$(rows.scala:36)                             
        at org.apache.spark.sql.catalyst.expressions.GenericInternalRow.isNullAt(rows.scala:195)
        at org.apache.spark.sql.catalyst.expressions.JoinedRow.isNullAt(JoinedRow.scala:70)
        ...
```

## How was this patch tested?

Added test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @HyukjinKwon @BryanCutler

**Comment by HyukjinKwon** [comment]
Cool @viirya! I will take a closer look within 2 days and get this in

**Comment by viirya** [comment]
Thanks! @HyukjinKwon

**Comment by viirya** [comment]
retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 1]
```diff
@@ -969,4 +970,26 @@ class JoinSuite extends QueryTest with SharedSQLContext {
           Seq(Row(0.0d, 0.0/0.0)))))
     }
   }
+
+  test("SPARK-28323: PythonUDF should be able to use in join condition") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: Nice. can we add an assume too? `assume(shouldTestScalarPandasUDFs)`

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: Ok.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala [hunk 2]
```diff
@@ -179,7 +179,7 @@ object ExtractPythonUDFs extends Rule[LogicalPlan] with PredicateHelper {
             validUdfs.forall(PythonUDF.isScalarPythonUDF),
             "Can only extract scalar vectorized udf or sql batch udf")
 
-          val resultAttrs = udfs.zipWithIndex.map { case (u, i) =>
+          val resultAttrs = validUdfs.zipWithIndex.map { case (u, i) =>
```
Comment (BryanCutler) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]: oof, nice catch!

**Review by BryanCutler — APPROVED** [review]
LGTM

**Comment by BryanCutler** [comment]
merged to master, thanks @viirya !

**Comment by BryanCutler** [comment]
jira seems down, so I wasn't able to resolve the issue. will try later

**Comment by HyukjinKwon** [comment]
Yea, LGTM too!

**Comment by viirya** [comment]
Thanks! @HyukjinKwon @BryanCutler

**Pull request closed without merging**